### ISSUE TITLE
Support a way to load input data from AWS S3 (or a recipe for how to do it)

### ISSUE THREAD

**Issue by dgoldenberg-audiomack** [tensorflow/io#1252]
I'm not sure if #27 is related; there's something about S3 there.

Environment:
- AWS EMR emr-6.2.0
- Spark 3.0.1
- TF 2.4.0
- TF IO 0.17.0

My use-case is, I have Parquet files stored in AWS S3 and I want to load them into a TF dataset. So, basically:

```
filepath = "s3://my-bucket/dir1/dir2/part-....etc.parquet"
common_ds = tfio.IODataset.from_parquet(filepath, columns=columns)
```

Currently, I'm getting a error as follows:

>   File "/home/hadoop/.local/lib/python3.7/site-packages/tensorflow_io/core/python/ops/io_dataset.py", line 275, in from_parquet
    filename, columns=columns, internal=True
  File "/home/hadoop/.local/lib/python3.7/site-packages/tensorflow_io/core/python/ops/parquet_dataset_ops.py", line 31, in __init__
    filename, shared=filename, container="ParquetIODataset"
  File "<string>", line 7960, in io_parquet_readable_info
  File "<string>", line 8009, in io_parquet_readable_info_eager_fallback
  File "/home/hadoop/.local/lib/python3.7/site-packages/tensorflow/python/eager/execute.py", line 60, in quick_execute
    inputs, attrs, num_outputs)
tensorflow.python.framework.errors_impl.NotFoundError: Object s3://my-bucket/emr/spark_apps/my-app/part-00000-5a356839-2acb-489d-aa57-c97dbc90e6f9-c000.snappy.parquet does not exist [Op:IO>ParquetReadableInfo]

The file actually exists; it looks like TF IO can't read out of S3.

Is the ability actually supported in TF IO?  If so, how does one do that?  If not, the request is to add support for it.

**Comment by kvignesh1420** [comment]
@dgoldenberg-audiomack we are currently in the process of migrating filesystem plugins from `tensorflow/tensorflow` to `tensorflow/io`.  Thus, we are currently using a different `scheme` in `tensorflow/io` called `s3e` to connect to s3 storage so that it doesn't conflict with the existing symbols in tensorflow. 

So, as a sanity check for reading parquet files as an `IODataset`, can you check whether you are able to directly read one of the file using:

```
import tensorflow as tf
import tensorflow_io as tfio

data = tf.io.read_file("s3e://{}/{}".format(bucket_name, key_name))
# change the path as necessary
```

Please let me know.

**Comment by dgoldenberg-audiomack** [comment]
Hi @kvignesh1420, thanks for your quick response.

I've changed the code to use the "s3e" protocol. I still got an error, pls see below. Just checked; the file does exist.

- Is there anything special I need to do to enable the s3e?
- I see [this writeup](https://github.com/tensorflow/examples/blob/master/community/en/docs/deploy/s3.md) on TF on S3. Safe to assume this isn't relevant since my code is running in EMR, and should be able to run under the right policy and have access? Since I'm able to write the data, I should be able to read it (into TF).
- I'm wondering how relevant [EMRFS](https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-fs.html) is, however, rather than try and bring that in, I'm hoping we can get TF IO to just read the data

> 2020-12-30 13:45:32.167646: W tensorflow/core/framework/op_kernel.cc:1763] OP_REQUIRES failed at io_kernel.h:71 : Not found: Object s3e://my-bucket/emr/spark_apps/movielens-100k-tf/part-00000-8675ebba-3316-4cc0-ba4b-9aa98c7315f2-c000.snappy.parquet does not exist
Traceback (most recent call last):
  File "/mnt/tmp/spark-e9875a06-ba1a-4995-bc32-f8cef31fe0bc/recsys_tfrs_proto.py", line 308, in <module>
    main(sys.argv)
  File "/home/hadoop/.local/lib/python3.7/site-packages/tensorflow/python/autograph/impl/api.py", line 620, in wrapper
    return func(*args, **kwargs)
  File "/mnt/tmp/spark-e9875a06-ba1a-4995-bc32-f8cef31fe0bc/recsys_tfrs_proto.py", line 55, in main
    movies_ds, ratings_ds = load_data(spark)
  File "/home/hadoop/.local/lib/python3.7/site-packages/tensorflow/python/autograph/impl/api.py", line 620, in wrapper
    return func(*args, **kwargs)
  File "/mnt/tmp/spark-e9875a06-ba1a-4995-bc32-f8cef31fe0bc/recsys_tfrs_proto.py", line 278, in load_data
    movies_ds = load_dataset(movies_filenames, movies_columns)
  File "/mnt/tmp/spark-e9875a06-ba1a-4995-bc32-f8cef31fe0bc/recsys_tfrs_proto.py", line 185, in load_dataset
    common_ds = tfio.IODataset.from_parquet(files[0], columns=columns)
  File "/home/hadoop/.local/lib/python3.7/site-packages/tensorflow_io/core/python/ops/io_dataset.py", line 275, in from_parquet
    filename, columns=columns, internal=True
  File "/home/hadoop/.local/lib/python3.7/site-packages/tensorflow_io/core/python/ops/parquet_dataset_ops.py", line 31, in __init__
    filename, shared=filename, container="ParquetIODataset"
  File "<string>", line 7960, in io_parquet_readable_info
  File "<string>", line 8009, in io_parquet_readable_info_eager_fallback
  File "/home/hadoop/.local/lib/python3.7/site-packages/tensorflow/python/eager/execute.py", line 60, in quick_execute
    inputs, attrs, num_outputs)
tensorflow.python.framework.errors_impl.NotFoundError: Object s3e://my-bucket/emr/spark_apps/movielens-100k-tf/part-00000-8675ebba-3316-4cc0-ba4b-9aa98c7315f2-c000.snappy.parquet does not exist [Op:IO>ParquetReadableInfo]

**Comment by dgoldenberg-audiomack** [comment]
@kvignesh1420,

Following up on my previous comment,

In `tests/test_s3_eager.py` in TF IO v.0.17.0, I see the following:

```
os.environ["AWS_REGION"] = "us-east-1"
os.environ["AWS_ACCESS_KEY_ID"] = "ACCESS_KEY"
os.environ["AWS_SECRET_ACCESS_KEY"] = "SECRET_KEY"

os.environ["S3_ENDPOINT"] = "localhost:4566"
os.environ["S3_USE_HTTPS"] = "0"
os.environ["S3_VERIFY_SSL"] = "0"

# TODO: The following is not working yet, need update to use
# s3 implementation with module file system
content = tf.io.read_file("s3e://{}/{}".format(bucket_name, key_name))
```

Which raises two questions:
1. Is s3e not actually enabled/not working yet? If so, when might it become enabled? Or will it be superseded by a full support for s3?
2. I see all the env var setups. Safe to assume those won't be necessary if running in EMR? I mean, PySpark is 'simply' able to write Parquet into S3 opaquely; it doesn't require a setup of these vars. In a Production setting, asking a job to know such creds as AWS_SECRET_ACCESS_KEY will sure raise eyebrows of our sysadmins, I suspect.

**Comment by kvignesh1420** [comment]
@dgoldenberg-audiomack 
1. `s3e` works as of now. The tests are passing in our Github Actions jobs. That TODO note is an old one and will be removed once the migration is complete.
2. I think you need to set the necessary env vars before reading the files from `s3` as of now. I am not sure of an alternative approach. Maybe @vnvo2409 @yongtang might be able to shed some more light on this.

**Comment by vnvo2409** [comment]
> Is `s3e` not actually enabled/not working yet?

`s3e` is working now ( at least for reading/writing ). I think we add that TODO because `s3e` isn't fully tested.

> Or will it be superseded by a full support for s3?

It will replace `s3` in TF 2.5.

@dgoldenberg-audiomack Could you set additional envs ( https://github.com/tensorflow/tensorflow/issues/44564#issuecomment-747599861 ) and we might get some insight about this issue ? Please remove sensitive information before posting the log !

**Comment by dgoldenberg-audiomack** [comment]
Hi @vnvo2409, @kvignesh1420,

Thanks for your comments.

I'll see if I can set up a test with the env vars in place. This alone poses a big issue due to the sensitive nature of the credentials. I've made this point earlier: "PySpark is 'simply' able to write Parquet into S3 opaquely; it doesn't require a setup of these vars. In a Production setting, asking a job to know such creds as AWS_SECRET_ACCESS_KEY will sure raise eyebrows of our sysadmins".

Can we have an enhancement request for the framework to allow for a seamless execution where, if you're within a security context like an AWS role, things "just work"?  Again, how is PySpark able to write to S3 w/o any creds fed into it, but TF is not able to read from it?

**Comment by kvignesh1420** [comment]
@dgoldenberg-audiomack to connect to `s3` using `pyspark` you can set the creds either in the spark context, as env vars or in the `spark-env` file I guess. Depends on your environment setup.
Reference: https://stackoverflow.com/questions/32155617/connect-to-s3-data-from-pyspark

Coming to the enhancement request, loading configurations for file-systems based on security contexts (other than env vars) is something that we are yet to look into.

**Comment by vnvo2409** [comment]
@dgoldenberg-audiomack 

I am thinking that your problem has nothing to do with the env and credentials. Because it throws `NotFoundError` rather than `FailedPreconditionError`. 

https://github.com/tensorflow/tensorflow/blob/9cfc4f740fadf04ce194e7e9b65decc0736503e4/tensorflow/core/platform/s3/s3_file_system.cc#L767-L769

So I think some additional logs could help us understand better what `aws-sdk-cpp` was doing behind the sence.

**Comment by dgoldenberg-audiomack** [comment]
@kvignesh1420 Explicitly handling creds of an environment such as e.g. production in code is something I suspect a lot of users of this functionality will have a problem with. Is there a ticket for "loading configurations for file-systems based on security contexts (other than env vars)"?  If not, I propose we create one, as an enhancement request.

**Comment by dgoldenberg-audiomack** [comment]
@vnvo2409 Might this be due to the fact that I'm reading the files immediately after they're created in S3? There may be some latency due to the 'eventual consistency'. If so, this won't his us in a prod scenario, only in this prototype scenario.  I'm going to try and put a delay between the writing and the reading to see if it helps, while reading with s3e.

**Comment by vnvo2409** [comment]
> Explicitly handling creds of an environment such as e.g. production in code is something I suspect a lot of users of this functionality will have a problem with. Is there a ticket for "loading configurations for file-systems based on security contexts (other than env vars)"? If not, I propose we create one, as an enhancement request.

@dgoldenberg-audiomack 

We have an issue here https://github.com/tensorflow/tensorflow/issues/43344. The main blocker is `aws-sdk-cpp` seems  to decide that this feature is not high-priority and we use `aws-sdk-cpp` under the hook so it is really hard for getting this work properly.

**Comment by dgoldenberg-audiomack** [comment]
@vnvo2409 Hi, I've added my "vote"/comment on that issue (43344). IMO, it really should be a higher priority.

@kvignesh1420:
> we are currently using a different scheme in tensorflow/io called s3e to connect to s3 storage so that it doesn't conflict with the existing symbols in tensorflow.

The "s3e" scheme is not working for me. Please take a look at my isolated code sample (attached). All it's trying to do is load a single Parquet file (attached) out of S3 into a IODataset. 

Stack trace is below.

I had thought that I've got a latency due to the 'eventual consistency' in S3 since I was writing Parquet and reading it immediately after. Well, now my tester simply tries to load an existing Parquet file, so I've ruled out the possibility of an 'eventual consistency' issue.

Any ideas as to how to get the "s3e" going?  Any fix or workaround? Thanks.

Stack trace:

> 2021-01-05 16:26:43.186591: W tensorflow/core/framework/op_kernel.cc:1763] OP_REQUIRES failed at io_kernel.h:71 : Not found: Object s3e://my-bucket/emr/spark_apps/movielens-example/movies-parquet/part-00000-bade9597-4bf2-41ed-af16-3aad37559436-c000.snappy.parquet does not exist
Traceback (most recent call last):
  File "/mnt/tmp/spark-84d3b03c-41b6-4c48-b335-5f1d59091653/tf_io_1252_load_from_s3.py", line 33, in <module>
    main(sys.argv)
  File "/mnt/tmp/spark-84d3b03c-41b6-4c48-b335-5f1d59091653/tf_io_1252_load_from_s3.py", line 21, in main
    ds = tfio.IODataset.from_parquet(FILEPATH, columns=movies_columns)
  File "/home/hadoop/.local/lib/python3.7/site-packages/tensorflow_io/core/python/ops/io_dataset.py", line 275, in from_parquet
    filename, columns=columns, internal=True
  File "/home/hadoop/.local/lib/python3.7/site-packages/tensorflow_io/core/python/ops/parquet_dataset_ops.py", line 31, in __init__
    filename, shared=filename, container="ParquetIODataset"
  File "<string>", line 7960, in io_parquet_readable_info
  File "<string>", line 8009, in io_parquet_readable_info_eager_fallback
  File "/home/hadoop/.local/lib/python3.7/site-packages/tensorflow/python/eager/execute.py", line 60, in quick_execute
    inputs, attrs, num_outputs)
tensorflow.python.framework.errors_impl.NotFoundError: Object s3e://my-bucket/emr/spark_apps/movielens-example/movies-parquet/part-00000-bade9597-4bf2-41ed-af16-3aad37559436-c000.snappy.parquet does not exist [Op:IO>ParquetReadableInfo]

[tf_io_1252.zip](https://github.com/tensorflow/io/files/5771445/tf_io_1252.zip)

**Comment by kvignesh1420** [comment]
@dgoldenberg-audiomack based on your script, here are the steps that I performed based on your data and script:

1. spin up a local instance of localstack that we use for CI testing:
```
 bash tests/test_aws/aws_test.sh
```

2. Upload the file that you provided into the s3 filesystem:
```python
import os
import boto3

os.environ["AWS_REGION"] = "us-east-1"
os.environ["AWS_ACCESS_KEY_ID"] = "ACCESS_KEY"
os.environ["AWS_SECRET_ACCESS_KEY"] = "SECRET_KEY"

BUCKET_NAME = "my-bucket"
S3_PARQUET_DIRPATH = "emr/spark_apps/movielens-example"
LOCAL_FILEPATH = "/Users/vignesh/Downloads/tf_io_1252/part-00000-597122cd-9302-4e04-a24d-8594f84bf61e-c000.snappy.parquet"
FILEPATH = f"s3e://{BUCKET_NAME}/{S3_PARQUET_DIRPATH}/movies-parquet/part-00000-bade9597-4bf2-41ed-af16-3aad37559436-c000.snappy.parquet"

client = boto3.client(
    "s3", region_name="us-east-1", endpoint_url="http://localhost:4566"
)
client.create_bucket(Bucket=BUCKET_NAME)

client.upload_file(LOCAL_FILEPATH, BUCKET_NAME, FILEPATH)
```

3. Read the parquet dataset
```python
In [1]: import os
In [2]: import tensorflow_io as tfio
In [3]: os.environ["S3_ENDPOINT"] = "localhost:4566"
   ...: os.environ["S3_USE_HTTPS"] = "0"
   ...: os.environ["S3_VERIFY_SSL"] = "0"

In [4]: FILEPATH = 's3e://my-bucket/emr/spark_apps/movielens-example/movies-parquet/part-00000-bade9597-4bf2-41ed-af16-3aad37559436-c000.snappy.parquet'
   ...:

In [5]: movies_columns = ["movie_id", "movie_title", "genres"]

In [6]: ds = tfio.IODataset.from_parquet(FILEPATH, columns=movies_columns)
2021-01-05 23:31:19.935645: I tensorflow_io/core/kernels/cpu_check.cc:128] Your CPU supports instructions that this TensorFlow IO binary was not compiled to use: SSE4.2 AVX AVX2 FMA
2021-01-05 23:31:20.065654: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-01-05 23:31:20.068382: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.

In [7]: it = iter(ds)
2021-01-05 23:31:27.617149: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)

In [8]: next(it)
Out[8]:
OrderedDict([('movie_id', <tf.Tensor: shape=(), dtype=int32, numpy=1>),
             ('movie_title',
              <tf.Tensor: shape=(), dtype=string, numpy=b'Toy Story (1995)'>),
             ('genres',
              <tf.Tensor: shape=(), dtype=string, numpy=b'Adventure|Animation|Children|Comedy|Fantasy'>)])
```

Observation:

```python
# Actual file name
part-00000-597122cd-9302-4e04-a24d-8594f84bf61e-c000.snappy.parquet"
# Filename in code
part-00000-bade9597-4bf2-41ed-af16-3aad37559436-c000.snappy.parquet"
```
The actual filename and the one mentioned in code is different. Is this something that you have handled elsewhere?
Also, as per @vnvo2409 the error says `NotFoundError` and not some connectivity error to S3 (which would result in `FailedPreconditionError `).

**Comment by dgoldenberg-audiomack** [comment]
@kvignesh1420 Sorry, I may have copy and pasted a wrong path while preparing the python snippet.

Interestingly, both the s3 and s3e protocols work. What was missing was this:

```
    os.environ["AWS_REGION"] = "us-east-1"
    os.environ["S3_ENDPOINT"] = "s3.us-east-1.amazonaws.com"
    os.environ["S3_USE_HTTPS"] = "0"
    os.environ["S3_VERIFY_SSL"] = "0"
```

This gets the code going actually. However, ideally, one wouldn't have to set this up when running in EMR.

**Comment by nwatab** [comment]
```py
import os
import tensorflow as tf
import tensorflow_io as tfio

tf.__version__  # '2.9.2'
tfio.__version__  # '0.28.0'

os.environ["AWS_REGION"] = "ap-northeast-1"
os.environ["AWS_ACCESS_KEY_ID"] = "***"
os.environ["AWS_SECRET_ACCESS_KEY"] = "***"
os.environ["S3_ENDPOINT"] = "s3.ap-northeast-1.amazonaws.com"
os.environ["S3_USE_HTTPS"] = "0"
os.environ["S3_VERIFY_SSL"] = "0"

tf.io.read_file('s3://mubucket/data.jpeg')
```

Not working.

```UnimplementedError: File system scheme 's3' not implemented (file: 's3://mybucket/data.jpeg') [Op:ReadFile]
